Partner Manager 2.16.0 (#200)
mulesoft/docs-partner-manager: modules/ROOT/pages/access-source-message-attributes.adoc

= Accessing Source Message Attributes in Translation Maps

> *Vijay, I am finding the text describing the use cases awkward when I use the term "accessing." To me, you are adding source attributes to translation maps or maybe exposing them, but not really accessing them. If you agree with "Exposing Source Message Attributes in Translation Maps,"" I will change the use case wording accordingly.* 

Reviewer: Hi Diane - I believe we are aligned on keeping this as "Accessing source message attributes" per our discussion earlier.

Author: OK

> When you create translation maps for inbound or outbound message flows, you can access source message attributes in those maps. The source attributes in the maps appear in the translated message payload, and you can view them on the activity screen for transmissions that use the map. 

Reviewer: Please change as follows:
When you create translation or reference maps for inbound or outbound message flows, you can access source message attributes in those maps. You can then map the attributes out in the translated message payload, or use the attribute values in your mapping logic to perform dynamic condition execution.

Author: OK - changed to this because "map the attributes out" doesn't sound right to me:

When you create translation or reference maps for inbound or outbound message flows, you can access source message attributes in those maps. You can then map the attributes to fields in the translated message payload or use the attribute values in your mapping logic to perform dynamic condition execution.

Some common use cases include:

* Achieving end-to-end tracing
+
Accessing transmission IDs in the translation maps enables you to achieve end-to-end tracing becuase these IDs are passed to the translated JSON or XML output file.
+
* Mapping attributes to pass them to the backend
+

> For example, if the backend must see the source filename or a specific HTTP header, you can access the filename or HTTP header in the translation map so that it appears in the translated data payload.

Reviewer: Please change as follows:

For example, if the backend must see the source filename or a specific HTTP header, you can access the filename or HTTP header and map the values to the translated message payload fields.

Author: OK

+
* Applying mapping logic based either on a filename part or pattern within the transformation
+
For example, you can access a filename prefix through substring operations and can build dynamic mapping logic based on this prefix. 
+
* Mapping attributes to make them available in the search
+

> For example, you can access the source filename or a specific HTTP header in the translation map so that you can easily search for these values in the Activity screens.

Reviewer: Please change as follows:
For example, you can access the source filename or a specific HTTP header value as a custom message attribute in in the message type's reference map so that you can easily search for these values in the Activity screens.

Reviewer: I would also move the "Some common use cases" section to appear after the section that explains what fields are available in the b2bMessageContext variable - as we are referring to things like transmission id, file name etc - which are explained better in the below section.

The source attributes that you can access are available within a predefined variable named `b2bMessageContext`. The content of this variable varies based on the source endpoint's protocol.

== AS2 Source Endpoint

Messages received via AS2 contain the following fields in the `b2bMessageContext` variable:

* sourceProtocol
* transmissionId
* sourceEndpointAttributes:
** as2MessageId
** as2From
** as2To
** fileName
** subject
** fileSize 
* customProtocolHeaders
+

> These headers always begin with the prefix `x-apm-`.

Reviewer: Please change as follows:
For security reasons, only the headers with the prefix 'x-apm-' are made available in the 'b2bMessageContext' variable.

  }
}
----

== FTP and SFTP Source Endpoint
 
Messages received via FTP or SFTP contain the following fields in the `b2bMessageContext` variable:

* sourceProtocol
* transmissionId
* sourceEndpointAttributes:
** fileName
** filePath
** fileSize
** fileTimeStamp
** partnerReferenceId
+
This attribute is available only for Receive from Partner endpoints that have the partner reference ID configured.
+
* messageId

For example:

[source,json]
----
{
  "b2bMessageContext": {
    "sourceProtocol": "SFTP",
    "transmissionId": "3ef5654a-5cda-41b1-8909-59a28c8b1537",
    "sourceEndpointAttributes": {
      "fileName": "RandomMessageType_PO_20230331_1559_003.xml",
      "filePath": "/partners/pacific-corp/in-orders",
      "fileSize": 4889,
      "fileTimeStamp": "2023-04-05T15:16:29",
      "partnerReferenceId": "PACIFICCORP"
    },
    "messageId": "cdaf6c12-bb3b-4978-9ac6-e8ace053f7c5"
  }
}
----

== HTTP and HTTPS Source Endpoint

Messages received via HTTP or HTTPS contain the following fields in the `b2bMessageContext` variable:

* sourceProtocol
* transmissionId
* sourceEndpointAttributes:
** path
** fileSize
* customProtocolHeaders
+

> These headers always begin with the prefix `x-apm-`.

Reviewer: Please change as follows:
For security reasons, only the headers with the prefix 'x-apm-' are made available in the 'b2bMessageContext' variable.

= Accessing Source Message Attributes

When you create translation or reference maps for inbound or outbound message flows, you can access source message attributes in those maps. You can then map the attributes to fields in the translated message payload or use the attribute values in your mapping logic to perform dynamic condition execution.

== Common Use Cases

Accessing source message attributes helps in many use cases, such as with performing dynamic message routing decisions, performing specific functions, and mapping fields within the data transformation in the message flow execution.

Some common use cases include:

> * Achieving end-to-end tracing

Reviewer 2: Is achieving tracing the outcome? Or something like "Tracing message routing end-to-end"?

+

> Accessing transmission IDs in the translation maps enables you to achieve end-to-end tracing because these IDs are passed to the translated JSON or XML output file.

Reviewer 2: ```suggestion
Accessing transmission IDs in the translation maps enables end-to-end tracing because these IDs are passed to the translated JSON or XML output file.
```

+

> * Mapping attributes to pass them to the back end

Reviewer 3: ```suggestion
* Mapping attributes to pass them to the backend
```

+

> For example, if the backend must see the source filename or a specific HTTP header, you can access the filename or HTTP header and map the values to the translated message payload fields.

Reviewer 2: I keep wanting to hyphenate these compound adjectives.

Reviewer 2: Backend OK as a noun in this case?

> For example, if the back end must see the source filename or a specific HTTP header, you can access the filename or HTTP header and map the values to the translated message payload fields.

Reviewer 3: ```suggestion
For example, if the backend must see the source filename or a specific HTTP header, you can access the filename or HTTP header and map the values to the translated message payload fields.
```

+
* Applying mapping logic based either on a filename part or pattern within the transformation
+

> For example, you can access a filename prefix through substring operations and can build dynamic mapping logic based on this prefix: 

Reviewer 2: ```suggestion
For example, you can access a filename prefix through substring operations and then build dynamic mapping logic based on this prefix.
```

Reviewer 2: Unless the following bullets are a sublist?

+
* Mapping attributes to make them available in the search
+

> For example, you can access the source filename or a specific HTTP header value as a custom message attribute in in the message type's reference map so that you can easily search for these values in the Activity screens.

Reviewer 2: ```suggestion
For example, you can access the source filename or a specific HTTP header value as a custom message attribute in the message type's reference map to easily search for these values in the Activity screens.
```

> For example, you can access the source filename or a specific HTTP header value as a custom message attribute in in the message type's reference map to easily search for these values in the Activity screens.

Reviewer 3: ```suggestion
For example, you can access the source filename or a specific HTTP header value as a custom message attribute in the message type's reference map to easily search for these values in the Activity screens.
```

For more information, see <<implementation-examples>>.

> The source attributes that you can access are available within a predefined variable named `b2bMessageContext`. The content of this variable varies based on the source endpoint's protocol.

Reviewer 2: variable varies - maybe a different word here? differs? depends on?

== AS2 Source Endpoint

Messages received via AS2 contain the following fields in the `b2bMessageContext` variable:

* sourceProtocol
* transmissionId
* sourceEndpointAttributes:
** as2MessageId
** as2From
** as2To
** fileName
** subject
** fileSize 
* customProtocolHeaders
+

> For security reasons, only the headers with the prefix 'x-apm-' are available in the 'b2bMessageContext' variable.

Reviewer 2: ```suggestion
For security reasons, only the headers with the prefix `x-apm-` are available in the `b2bMessageContext` variable.
```

  }
}
----

== FTP and SFTP Source Endpoint
 
Messages received via FTP or SFTP contain the following fields in the `b2bMessageContext` variable:

* sourceProtocol
* transmissionId
* sourceEndpointAttributes:
** fileName
** filePath
** fileSize
** fileTimeStamp
** partnerReferenceId
+
This attribute is available only for Receive from Partner endpoints that have the partner reference ID configured.
+
* messageId

For example:

[source,json]
----
{
  "b2bMessageContext": {
    "sourceProtocol": "SFTP",
    "transmissionId": "3ef5654a-5cda-41b1-8909-59a28c8b1537",
    "sourceEndpointAttributes": {
      "fileName": "RandomMessageType_PO_20230331_1559_003.xml",
      "filePath": "/partners/pacific-corp/in-orders",
      "fileSize": 4889,
      "fileTimeStamp": "2023-04-05T15:16:29",
      "partnerReferenceId": "PACIFICCORP"
    },
    "messageId": "cdaf6c12-bb3b-4978-9ac6-e8ace053f7c5"
  }
}
----

== HTTP and HTTPS Source Endpoint

Messages received via HTTP or HTTPS contain the following fields in the `b2bMessageContext` variable:

* sourceProtocol
* transmissionId
* sourceEndpointAttributes:
** path
** fileSize
* customProtocolHeaders
+

> For security reasons, only the headers with the prefix 'x-apm-' are available in the 'b2bMessageContext' variable.

Reviewer 2: ```suggestion
For security reasons, only the headers with the prefix `x-apm-` are available in the `b2bMessageContext` variable.
```

* messageId   

For example:

[source,json]
----
{ 
  "b2bMessageContext": {
    "sourceProtocol": "HTTPS",
    "transmissionId": "a7a800ad-4117-40c1-98a0-79892c982fd4",
    "sourceEndpointAttributes": {
    "path": "/",
    "fileSize": 866
    },
    "customProtocolHeaders": {
      "x-apm-order-number": "ORD12345-A",
      "x-apm-order-type": "drop-ship"
    },  
    "messageId": "cdaf6c12-bb3b-4978-9ac6-e8ace053f7c5"
  }
}
----

[[implementation-examples]]
== Use Case Implementation Examples

The following examples show how to implement some common use cases for source message attributes.

> This example shows how to map the `messageId` and `fileName` attributes in the translated data payload within a message flow's translation map:

Reviewer 2: One thought to streamline and improve scannability is to remove the "This example shows how to" intro from each and make these headings or bullets.

			values: [payload.ns0#B2BMessage.ns0#Data.ns0#PurchaseOrder.ns0#PurchaseOrderNumber]
		},
		{
			alias: "buyerEmail",
			values: [payload.ns0#B2BMessage.ns0#Data.ns0#PurchaseOrder.ns0#BuyerEmailID]
		},
		{
			alias: "sourceFileName",
			values: [vars.b2bMessageContext.sourceEndpointAttributes.fileName default “UNKNOWN]
		}
	]
}
----

This example shows how to map a custom header on the received HTTP message as a custom message attribute:

----
%dw 2.0
output application/json
ns ns0 http://xmlns.mulesoft.com/enterpriseobjects/b2b/
---
{
	partnerReferenceId: payload.ns0#B2BMessage.ns0#Data.ns0#PurchaseOrder.ns0#Seller,
	hostReferenceId: payload.ns0#B2BMessage.ns0#Data.ns0#PurchaseOrder.ns0#Buyer,
	businessDocumentKey: payload.ns0#B2BMessage.ns0#Data.ns0#PurchaseOrder.ns0#PurchaseOrderNumber,
	customAttributes: [
		{
			alias: "purchaseOrderNumber", 
			values: payload.ns0#B2BMessage.ns0#Data.*ns0#PurchaseOrder map ( purchaseOrder , indexOfPurchaseOrder ) 
						-> purchaseOrder.ns0#PurchaseOrderNumber default "UNKNOWN"
		},
		{
			alias: "supplierName", 
			values: [vars.b2bMessageContext.customProtocolHeaders.'x-apm-supplierName' default "UNKNOWN"]
		}
	]
}
----

== See Also

> * xref:partner-manager-maps.adoc[Translation Maps]

Reviewer 2: If you leave out the title from the [ ], it gets picked up automatically and you don't have to maintain the xrefs if the titles change.
